---
title: "DEG"
author: "Debbie Leung"
date: "3/23/2021"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Installing Packages

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
```

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("limma")
```

```{r, eval = FALSE}
#install.packages("shiny")
```

```{r}
library(dplyr)
library(tidyr)
```

```{r}
#install.packages("htmltools")
library(htmltools)
#source("https://bioconductor.org/biocLite.R")
#biocLite("DESeq2")
```

```{r}
library( "DESeq2" )
library(ggplot2)
```


## Filtering RSEM table

```{r, eval = FALSE}
ann <- read.delim("ann_table.txt")
head(ann)

# Select only SeqName column
ann <- ann %>% pull(SeqName)
```

```{r, eval = FALSE}
rsem <- read.delim("Transcript_RSEM_count_table.txt")
result <- filter(rsem, Name %in% ann)
write.csv(result, "RSEM_filtered.csv")
```

## Read data

```{r}
countData <- read.csv("FinalRSEM_air.csv", header = TRUE, sep = ",")
head(countData)
```

```{r}
metaData <- read.csv('Metadata_air.csv', header = TRUE, sep = ",", stringsAsFactor=TRUE)
metaData
```
```{r}
dds <- DESeqDataSetFromMatrix(countData=countData, 
                              colData=metaData, 
                              design = ~ Rate + Tide + Time, tidy = TRUE)
```

```{r}
dds <- DESeq(dds)
#?DESeq
```

```{r}
res <- results(dds)
res <- results(dds, tidy=TRUE)
#write.csv(as.data.frame(res), file="air.csv")
```

```{r}
summary(res) 
```

```{r}
res <- res[order(res$padj),]
head(res)
```

```{r}
annotation <- read.csv("annotation.csv", header = TRUE, sep = ",", stringsAsFactor=TRUE)
```

### LRT Test

```{r}
dds_rate <- DESeq(dds, test="LRT", reduced=~Tide + Time)
res_rate <- results(dds_rate)
res_rate <- res_rate[order(res_rate$padj),]
res_rate
#summary(results_rate)
#clusters <- degPatterns(cluster_rlog, metadata = meta, time="time", col="treatment")
```
```{r}
summary(res_rate)
```


```{r}
dds_tide <- DESeq(dds, test="LRT", reduced=~Rate + Time)
res_tide <- results(dds_tide)
res_tide <- res_tide[order(res_tide$padj),]
head(res_tide)
```

```{r}
dds_time <- DESeq(dds, test="LRT", reduced=~Rate + Tide)
res <- results(dds_time)
```


## Volcano Plot
Reference: https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html
```{r}
#reset par
par(mfrow=c(1,1))
# Make a basic volcano plot
with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-3,3)))

# Add colored points: blue if padj<0.01, red if log2FC>1 and padj<0.05)
with(subset(res, padj<.01 ), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, padj<.01 & abs(log2FoldChange)>2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))

```

## Heatmap

Which transformation method should I use? Options: shifted logarithm transformation, the regularized log transformation and the variance stabilizing transformation

```{r}
write.csv(assay(ntd)[select,], file="dds_norm_counts.csv")
```


```{r}
ntd <- normTransform(dds)

#install.packages("pheatmap")
library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("Rate","Tide", "Time")])
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```
```{r}
vsdata <- vst(dds, blind=FALSE)
sampleDists <- dist(t(assay(vsdata)))
#install.packages("RColorBrewer")
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsdata$Group, vsdata$Tide, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

```{r}
# Gene clustering

#install.packages("gplots")
#library(DEFormats)
#library(gplots)
mcols(dds) <- cbind(mcols(dds), annotation$Gene)
all(rownames(dds) == annotation$Transcript)
sum(is.na(annotation$Gene))
sum(duplicated(annotation$Gene))
vsdata <- vst(dds, blind=FALSE)
heatmap.2(assay(vsdata)[select,], col = redgreen, Rowv = TRUE, Colv = FALSE, scale="none",dendrogram="none", trace="none", margin=c(5, 6), labRow = annotation[select,"Gene"])
heatmap.2(assay(vsdata)[select,], col = redgreen, Rowv = FALSE, Colv = FALSE, scale="none",dendrogram="none", trace="none", margin=c(10, 6), labRow = annotation[select,"Gene"])
```

```{r}
library("genefilter")
topVarGenes <- head(order(rowVars(assay(vsdata)), decreasing = TRUE), 20)
mat  <- assay(vsdata)[ topVarGenes, ]

# find gene names corresponding to transcripts
library(magrittr)
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsdata)[, c("Rate","Tide","Time")])
mat %>% set_rownames(annotation$Gene[ topVarGenes ]) -> mat
pheatmap(mat, annotation_col = anno)
```

## PCA
```{r}
#First we need to transform the raw count data
#vst function will perform variance stabilizing transformation

plotPCA(vsdata, intgroup="Rate")
plotPCA(vsdata, intgroup="Rate", "Tide", "Time")
```
```{r}
plotPCA(vsdata, intgroup=c("Rate", "Tide", "Time"))

pcaData <- plotPCA(vsdata, intgroup=c("Rate", "Tide", "Time"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Rate, fill=Tide, shape=Time)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() 
```
```{r}
head(vsdata)
```


```{r}
rld <- rlog(dds)
plotPCA(rld)

# also possible to perform custom transformation:
dds <- estimateSizeFactors(dds)
# shifted log of normalized counts
se <- SummarizedExperiment(log2(counts(dds, normalized=TRUE) + 1),
                           colData=colData(dds))
# the call to DESeqTransform() is needed to
# trigger our plotPCA method.
plotPCA( DESeqTransform( se ) )
```

```{r}
library(ggpubr)
library(factoextra)
```


## Venn Diagram

## Tutorials
- https://lashlock.github.io/compbio/R_presentation.html
- https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis 
